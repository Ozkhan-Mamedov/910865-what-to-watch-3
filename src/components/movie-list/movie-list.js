import React from "react";
import PropTypes from "prop-types";

import SmallMovieCard from "../small-movie-card/small-movie-card";

class MovieList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      activeCard: -1
    };

    this.cardHoverHandler = this.cardHoverHandler.bind(this);
  }

  cardHoverHandler(id) {
    if (id !== this.state.activeCard) {
      this.setState({
        activeCard: id,
      });
    }
  }

  render() {
    const {films, filmNameClickHandler} = this.props;

    return (
      <div className="catalog__movies-list">
        {
          films.map((film, index) =>
            <SmallMovieCard
              film={film}
              key={index}
              filmNameClickHandler={filmNameClickHandler}
              cardHoverHandler={this.cardHoverHandler}
            />
          )
        }
      </div>
    );
  }
}

MovieList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.exact({
    name: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    releaseDate: PropTypes.string.isRequired,
    ratingScore: PropTypes.number.isRequired,
    ratingsNumber: PropTypes.number.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.arrayOf(PropTypes.string),
    description: PropTypes.arrayOf(PropTypes.string),
    preview: PropTypes.string.isRequired,
  })),
  filmNameClickHandler: PropTypes.func.isRequired
};

export default MovieList;
