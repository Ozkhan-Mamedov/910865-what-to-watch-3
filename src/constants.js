const ALL_GENRES = `ALL_GENRES`;
const GENRES = {
  ALL_GENRES: `All genres`,
  COMEDY: `Comedies`,
  DOCUMENTARY: `Documentary`,
  DRAMA: `Dramas`,
  HORROR: `Horror`,
  KIDS_AND_FAMILY: `Kids & Family`,
  ROMANCE: `Romance`,
  SCI_FI: `Sci-Fi`,
  THRILLER: `Thrillers`,
};
const MAX_GENRE_NUMBER = 10;
const CHANGE_FILTER_BY_GENRE = `CHANGE_FILTER_BY_GENRE`;
const GET_FILMS_BY_GENRE = `GET_FILMS_BY_GENRE`;
const CHANGE_ACTIVE_CARD = `CHANGE_ACTIVE_CARD`;
const INCREMENT_CARDS_NUMBER = `INCREMENT_CARDS_NUMBER`;
const TABS_LIST = [`Overview`, `Details`, `Reviews`];
const TABS_KEYS = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};
const MINUTES_IN_HOUR = 60;
const MONTH_SUBSTR = {
  START: 4,
  LENGTH: 3
};
const DAY_SUBSTR = {
  START: 8,
  LENGTH: 2
};
const YEAR_SUBSTR = {
  START: 11,
  LENGTH: 4
};
const MONTH_KEYS = [
  {
    month: `January`,
    key: `Jan`,
  },
  {
    month: `February`,
    key: `Feb`,
  },
  {
    month: `March`,
    key: `Mar`,
  },
  {
    month: `April`,
    key: `Apr`,
  },
  {
    month: `May`,
    key: `May`,
  },
  {
    month: `June`,
    key: `Jun`,
  },
  {
    month: `Jule`,
    key: `Jul`,
  },
  {
    month: `August`,
    key: `Aug`,
  },
  {
    month: `September`,
    key: `Sep`,
  },
  {
    month: `October`,
    key: `Oct`,
  },
  {
    month: `November`,
    key: `Nov`,
  },
  {
    month: `December`,
    key: `Dec`,
  },
];
const MOVIE_LIST = `MOVIE_LIST`;
const MORE_LIKE_THIS_LIST = `MORE_LIKE_THIS_LIST`;
const MORE_LIKE_THIS_FILMS = 4;
const MAX_CARD_RENDER_NUMBER = 8;
const PREVIEW_PLAYER_PROPERTIES = {
  WIDTH: 280,
  HEIGHT: 175,
};


export {
  ALL_GENRES,
  CHANGE_FILTER_BY_GENRE,
  GET_FILMS_BY_GENRE,
  GENRES,
  MAX_GENRE_NUMBER,
  TABS_LIST,
  TABS_KEYS,
  MINUTES_IN_HOUR,
  CHANGE_ACTIVE_CARD,
  MONTH_SUBSTR,
  DAY_SUBSTR,
  YEAR_SUBSTR,
  MONTH_KEYS,
  MORE_LIKE_THIS_LIST,
  MOVIE_LIST,
  MORE_LIKE_THIS_FILMS,
  MAX_CARD_RENDER_NUMBER,
  INCREMENT_CARDS_NUMBER,
  PREVIEW_PLAYER_PROPERTIES,
};
