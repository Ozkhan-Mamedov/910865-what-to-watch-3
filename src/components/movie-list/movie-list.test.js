import React from "react";
import renderer from "react-test-renderer";

import {MovieList} from "./movie-list";

import {MORE_LIKE_THIS_LIST, MOVIE_LIST} from "../../constants";

const films = [
  {
    name: `Film#1`,
    picture: `picture#1`,
    genre: `genre#1`,
    id: 1,
    releaseDate: `2020`,
    ratingScore: 1,
    ratingsNumber: 2,
    director: `Director#1`,
    starring: [`Actor#1`, `Actor#2`, `Actor#3`, `Actor#4`],
    description: `Test paragraph #1.`,
    preview: `preview#1`,
    runTime: 130,
    previewImage: `preview image#1`,
    videoLink: `video link#1`,
    isFavorite: false,
    backgroundColor: `background color#1`,
    backgroundImage: `background image#1`,
  },
  {
    name: `Film#2`,
    picture: `picture#2`,
    genre: `genre#2`,
    id: 2,
    releaseDate: `2021`,
    ratingScore: 2,
    ratingsNumber: 3,
    director: `Director#2`,
    starring: [`Actor#5`, `Actor#6`, `Actor#7`, `Actor#8`],
    description: `Test paragraph #2.`,
    preview: `preview#2`,
    runTime: 131,
    previewImage: `preview image#2`,
    videoLink: `video link#2`,
    isFavorite: false,
    backgroundColor: `background color#2`,
    backgroundImage: `background image#2`,
  },
  {
    name: `Film#3`,
    picture: `picture#3`,
    genre: `genre#3`,
    id: 3,
    releaseDate: `2022`,
    ratingScore: 3,
    ratingsNumber: 4,
    director: `Director#2`,
    starring: [`Actor#9`, `Actor#10`, `Actor#11`, `Actor#12`],
    description: `Test paragraph #3.`,
    preview: `preview#3`,
    runTime: 132,
    previewImage: `preview image#3`,
    videoLink: `video link#3`,
    isFavorite: false,
    backgroundColor: `background color#3`,
    backgroundImage: `background image#3`,
  },
];

it(`MovieList component renders correctly for main page`, () => {
  const tree = renderer
    .create(<MovieList filteredFilmsList={films} films={films} filmNameClickHandler={() => {}} list={MOVIE_LIST}/>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`MovieList component renders correctly for details page`, () => {
  const tree = renderer
    .create(<MovieList films={films} filmNameClickHandler={() => {}} list={MORE_LIKE_THIS_LIST}/>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
