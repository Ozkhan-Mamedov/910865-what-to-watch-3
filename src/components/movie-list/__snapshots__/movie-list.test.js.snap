// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component renders correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Film#1"
        height="175"
        onClick={[Function]}
        src="picture#1"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
        onClick={[Function]}
      >
        Film#1
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Film#2"
        height="175"
        onClick={[Function]}
        src="picture#2"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
        onClick={[Function]}
      >
        Film#2
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Film#3"
        height="175"
        onClick={[Function]}
        src="picture#3"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
        onClick={[Function]}
      >
        Film#3
      </a>
    </h3>
  </article>
</div>
`;
